### ISSUE TITLE
PMA goes into infinite loop when flist has exactly one block and you allocate something of exactly that size

### ISSUE THREAD

**Issue by philipquirk** [urbit/ares#205]
I'm getting an infinite loop in the PMA.  Here's how it happens:

- state->flist has exactly one block.
- It allocates something which is the exact size of this block (in my case the size is just 1 page).
- in _flist_record_alloc it hits the [last case](https://github.com/urbit/ares/blob/status/rust/ares_pma/c-src/btree.c#L520). head is state->flist since there's only node and head->next is NULL and it frees head and sets state->flist to NULL.
- Now the next time the PMA tries to find a free block in _bt_falloc it fails to find one and calls _flist_grow to add one.
- _flist_grow calls ftruncate to increase the file size and calls _flist_insert to add the new memory to the flist.
- _flist_insert does not check if *dst is NULL and so the while loop immediately exists and it doesn't add any block.
- _bt_falloc tries again to find a free block and fails, so it calls _flist_grow and _flist_insert ot add more space, which doesn't check if *dst is NULL, ...

I tried a couple fixes for this such as making _flist_insert check if *dst is NULL and if so just allocating a new block, or doing this inside _flist_grow. Both cause crashes so I don't know what the real solution is.

**Comment by barter-simsum** [comment]
I had originally thought the partition striping branch contained a fix for this, but it doesn't seem to. While I work on a fix, you can unblock yourself by increasing `PMA_GROW_SIZE_p` to something much larger (to prevent the need to grow the persistent file at all). I spoke with @matthew-levan and managed to unblock him with this too.

Try multiplying the `PMA_GROW_SIZE_p` value by 10x (160M) or 100x (1.6G).

**Comment by barter-simsum** [comment]
actually, I believe the above PR should resolve this issue

**Comment by philipquirk** [comment]
This PR crashes for me with this error:

```
ares(82411,0x1d821dec0) malloc: *** error for object 0x600001fdc000: pointer being freed was not allocated
ares(82411,0x1d821dec0) malloc: *** set a breakpoint in malloc_error_break to debug
```

I had tried this myself but _flist_insert is also called on state->pending_flist which is sometimes NULL. So maybe that's part of it.

I also tried to do this in _flist_grow by checking if state->flist is NULL after the call to _flist_insert and I get a different crash. I have no idea why this doesn't fix it.
